pkg/cloudprovider/provider/vsphere/provider.go

```diff
@@ -385,6 +385,14 @@ func (p *provider) Create(machine *v1alpha1.Machine, _ *cloud.MachineCreateDelet
 		}()
 
 		if err := uploadAndAttachISO(ctx, finder, virtualMachine, localUserdataIsoFilePath, config.Datastore); err != nil {
+			// Destroy VM to avoid a leftover.
+			destroyTask, err := virtualMachine.Destroy(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("failed to destroy vm %s: %v", virtualMachine.Name(), err)
```

Comment:
Can you for both possible errors please also add the info that the uploadAndAttachIso failed and the error that happened there?